[jdbc-ping] SQLIntegrityConstraintViolationException: Duplicate entry

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

infinispan

### Describe the bug

I run into this issue when my database gets reprovisioned while the Keycloak server is still running.

The DB becomes unreachable for 15-30 seconds, and then it should be fine again. I don't understand why this is happening, and it's making me hesitant to upgrade. I was previously using `KC_CACHE_STACK: "kubernetes"`.

Couldn't this same state occur during normal operation—for example, if the DB is temporarily unreachable, or only one node can reach it, and so on?

If I stop the server and provision the DB, it works every time, and the `JGROUPS_PING` table appears to get cleaned up.

Maybe this is a very edge-case scenario and shouldn't be handled this way. I just want to confirm if it's 100% due to the moment when the DB connection becomes possible but the schema isn't fully there yet.

Shouldn`t it be self-healing?

```
2025-11-13 10:54:21,136 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) Error details:: org.keycloak.models.ModelDuplicateException: Duplicate resource error
        at org.keycloak.connections.jpa.support.EntityManagerProxy.convert(EntityManagerProxy.java:144)
        at org.keycloak.connections.jpa.support.EntityManagerProxy.invoke(EntityManagerProxy.java:112)
        at jdk.proxy2/jdk.proxy2.$Proxy57.runWithConnection(Unknown Source)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.prepareJGroupsAddress(JGroupsConfigurator.java:277)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.lambda$configureDiscovery$2(JGroupsConfigurator.java:249)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:461)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:439)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.lambda$configureDiscovery$3(JGroupsConfigurator.java:248)
        at org.keycloak.common.util.Retry.call(Retry.java:149)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.configureDiscovery(JGroupsConfigurator.java:248)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.configureJGroups(JGroupsConfigurator.java:147)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.lambda$singleSiteConfiguration$2(DefaultCacheEmbeddedConfigProviderFactory.java:229)
        at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:347)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:461)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:346)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:336)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.singleSiteConfiguration(DefaultCacheEmbeddedConfigProviderFactory.java:229)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.configureSingleSiteWithPersistentSessions(DefaultCacheEmbeddedConfigProviderFactory.java:178)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.createConfiguration(DefaultCacheEmbeddedConfigProviderFactory.java:166)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.lazyInit(DefaultCacheEmbeddedConfigProviderFactory.java:153)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.postInit(DefaultCacheEmbeddedConfigProviderFactory.java:102)
        at org.keycloak.services.DefaultKeycloakSessionFactory.initializeProviders(DefaultKeycloakSessionFactory.java:167)
        at org.keycloak.services.DefaultKeycloakSessionFactory.initializeProviders(DefaultKeycloakSessionFactory.java:164)
        at org.keycloak.services.DefaultKeycloakSessionFactory.initializeProviders(DefaultKeycloakSessionFactory.java:164)
        at org.keycloak.services.DefaultKeycloakSessionFactory.initProviderFactories(DefaultKeycloakSessionFactory.java:141)
        at org.keycloak.services.DefaultKeycloakSessionFactory.initProviderFactories(DefaultKeycloakSessionFactory.java:125)
        at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.init(QuarkusKeycloakSessionFactory.java:87)
        at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.createSessionFactory(QuarkusKeycloakApplication.java:60)
        at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:75)
        at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.onStartupEvent(QuarkusKeycloakApplication.java:50)
        at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication_Observer_onStartupEvent_VUhusbW4xizn9A5dimKJUcfvjHs.notify(Unknown Source)
        at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:365)
        at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:347)
        at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:81)
        at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:163)
        at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:114)
        at io.quarkus.runner.recorded.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
        at io.quarkus.runner.recorded.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:80)
        at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:139)
        at org.keycloak.quarkus.runtime.cli.Picocli.start(Picocli.java:913)
        at org.keycloak.quarkus.runtime.cli.command.AbstractAutoBuildCommand.runCommand(AbstractAutoBuildCommand.java:134)
        at org.keycloak.quarkus.runtime.cli.command.AbstractCommand.lambda$call$1(AbstractCommand.java:73)
        at java.base/java.util.Optional.orElseGet(Optional.java:364)
        at org.keycloak.quarkus.runtime.cli.command.AbstractCommand.call(AbstractCommand.java:72)
        at org.keycloak.quarkus.runtime.cli.command.AbstractCommand.call(AbstractCommand.java:37)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2031)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2469)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2461)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2423)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2425)
        at picocli.CommandLine.execute(CommandLine.java:2174)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:118)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:109)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:83)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:69)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:37)
Caused by: java.sql.SQLIntegrityConstraintViolationException: Duplicate entry 'uuid://00000000-0000-0000-0000-000000000003' for key 'JGROUPS_PING.PRIMARY'
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:118)
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:912)
        at com.mysql.cj.jdbc.ClientPreparedStatement.execute(ClientPreparedStatement.java:354)
        at io.agroal.pool.wrapper.PreparedStatementWrapper.execute(PreparedStatementWrapper.java:301)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.lambda$prepareJGroupsAddress$5(JGroupsConfigurator.java:285)
        at org.hibernate.engine.spi.SessionImplementor.lambda$runWithConnection$0(SessionImplementor.java:93)
        at org.hibernate.jdbc.WorkExecutor.executeWork(WorkExecutor.java:37)
        at org.hibernate.internal.AbstractSharedSessionContract.lambda$doWork$8(AbstractSharedSessionContract.java:1151)
        at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.coordinateWork(JdbcCoordinatorImpl.java:314)
        at org.hibernate.internal.AbstractSharedSessionContract.doWork(AbstractSharedSessionContract.java:1162)
        at org.hibernate.internal.AbstractSharedSessionContract.doWork(AbstractSharedSessionContract.java:1150)
        at org.hibernate.engine.spi.SessionImplementor.runWithConnection(SessionImplementor.java:90)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.keycloak.connections.jpa.support.EntityManagerProxy.invoke(EntityManagerProxy.java:100)
        ... 60 more
```


 

### Version

26.4.5

### Regression

- [ ] The issue is a regression

### Expected behavior

self recover from any state

### Actual behavior

not starting at all and keeps restarting with the error

### How to Reproduce?

provision db with a backup while server is running

### Anything else?

_No response_

Hi @pruivo,  provision db with a backup means. You delete the schema and recreate it from a backup. We do this for our dev instances for example. 


I just reproduced it like this.

Step 1: While Keycloak is running, delete the schema, restore I from any backup.
It will only log a message and continue to run afterwards, as soon as the schema is restored.



```
025-11-14 08:19:39,067 WARN  [org.keycloak.jgroups.protocol.KEYCLOAK_JDBC_PING2] (blocking-thread--p3-t2) Failed to fetch the cluster members from the database.: java.sql.SQLSyntaxErrorException: Table 'keycloak_dev.JGROUPS_PING' doesn't exist
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:121)
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:912)
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeQuery(ClientPreparedStatement.java:968)
        at io.agroal.pool.wrapper.PreparedStatementWrapper.executeQuery(PreparedStatementWrapper.java:80)
        at org.jgroups.protocols.JDBC_PING2.readFromDB(JDBC_PING2.java:230)
        at org.keycloak.jgroups.protocol.KEYCLOAK_JDBC_PING2.healthStatus(KEYCLOAK_JDBC_PING2.java:270)
        at org.keycloak.infinispan.health.impl.JdbcPingClusterHealthImpl.checkHealth(JdbcPingClusterHealthImpl.java:90)
        at org.keycloak.infinispan.health.impl.JdbcPingClusterHealthImpl$HealthRunner.lambda$trigger$0(JdbcPingClusterHealthImpl.java:138)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
        at java.base/java.lang.Thread.run(Thread.java:1583)

2025-11-14 08:19:49,050 WARN  [org.keycloak.infinispan.health.impl.JdbcPingClusterHealthImpl] (blocking-thread--p3-t2) Unable to check the cluster health because no coordinator has been found.
2025-11-14 08:19:59,049 INFO  [io.smallrye.health] (vert.x-eventloop-thread-0) SRHCK01001: Reporting health down status: {"status":"DOWN","checks":[{"name":"Keycloak cluster health check","status":"DOWN","data":{"Failing since":"2025-11-14 08:19:59,046"}},{"name":"Keycloak database connections async health check","status":"UP"}]}
2025-11-14 08:19:59,050 WARN  [org.keycloak.infinispan.health.impl.JdbcPingClusterHealthImpl] (blocking-thread--p3-t2) Unable to check the cluster health because no coordinator has been found.
2025-11-14 08:20:09,045 INFO  [io.smallrye.health] (vert.x-eventloop-thread-0) SRHCK01001: Reporting health down status: {"status":"DOWN","checks":[{"name":"Keycloak cluster health check","status":"DOWN","data":{"Failing since":"2025-11-14 08:19:59,046"}},{"name":"Keycloak database connections async health check","status":"UP"}]}
2025-11-14 08:20:09,049 WARN  [org.keycloak.infinispan.health.impl.JdbcPingClusterHealthImpl] (blocking-thread--p3-t2) Unable to check the cluster health because no coordinator has been found.
2025-11-14 08:20:19,045 INFO  [io.smallrye.health] (vert.x-eventloop-thread-0) SRHCK01001: Reporting health down status: {"status":"DOWN","checks":[{"name":"Keycloak cluster health check","status":"DOWN","data":{"Failing since":"2025-11-14 08:19:59,046"}},{"name":"Keycloak database connections async health check","status":"UP"}]}
2025-11-14 08:20:19,047 INFO  [io.smallrye.health] (vert.x-eventloop-thread-0) SRHCK01001: Reporting health down status: {"status":"DOWN","checks":[{"name":"Keycloak cluster health check","status":"DOWN","data":{"Failing since":"2025-11-14 08:19:59,046"}},{"name":"Keycloak database connections async health check","status":"UP"}]}
```



Step 2: Redeploy Keycloak in Kubernetes, I do this with a the code centric helm chart, but I guess this does not matter. Now the pods will not start anymore.


```
2025-11-14 08:42:52,452 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
2025-11-14 08:42:52,452 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) Error details:: org.keycloak.models.ModelDuplicateException: Duplicate resource error
        at org.keycloak.connections.jpa.support.EntityManagerProxy.convert(EntityManagerProxy.java:144)
        at org.keycloak.connections.jpa.support.EntityManagerProxy.invoke(EntityManagerProxy.java:112)
        at jdk.proxy2/jdk.proxy2.$Proxy57.runWithConnection(Unknown Source)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.prepareJGroupsAddress(JGroupsConfigurator.java:277)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.lambda$configureDiscovery$2(JGroupsConfigurator.java:249)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:461)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:439)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.lambda$configureDiscovery$3(JGroupsConfigurator.java:248)
        at org.keycloak.common.util.Retry.call(Retry.java:149)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.configureDiscovery(JGroupsConfigurator.java:248)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.configureJGroups(JGroupsConfigurator.java:147)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.lambda$singleSiteConfiguration$2(DefaultCacheEmbeddedConfigProviderFactory.java:229)
        at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:347)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:461)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:346)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:336)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.singleSiteConfiguration(DefaultCacheEmbeddedConfigProviderFactory.java:229)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.configureSingleSiteWithPersistentSessions(DefaultCacheEmbeddedConfigProviderFactory.java:178)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.createConfiguration(DefaultCacheEmbeddedConfigProviderFactory.java:166)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.lazyInit(DefaultCacheEmbeddedConfigProviderFactory.java:153)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.create(DefaultCacheEmbeddedConfigProviderFactory.java:91)
        at org.keycloak.spi.infinispan.impl.embedded.DefaultCacheEmbeddedConfigProviderFactory.create(DefaultCacheEmbeddedConfigProviderFactory.java:69)
        at org.keycloak.services.DefaultKeycloakSession.getOrCreateProvider(DefaultKeycloakSession.java:183)
        at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:172)
        at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.createEmbeddedCacheManager(DefaultInfinispanConnectionProviderFactory.java:185)
        at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.lazyInit(DefaultInfinispanConnectionProviderFactory.java:169)
        at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.create(DefaultInfinispanConnectionProviderFactory.java:101)
        at org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory.create(DefaultInfinispanConnectionProviderFactory.java:87)
        at org.keycloak.services.DefaultKeycloakSession.getOrCreateProvider(DefaultKeycloakSession.java:183)
        at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:172)
        at org.keycloak.models.sessions.infinispan.PersistentUserSessionProvider.lambda$migrate$13(PersistentUserSessionProvider.java:852)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
        at org.keycloak.models.sessions.infinispan.PersistentUserSessionProvider.migrate(PersistentUserSessionProvider.java:861)
        at org.keycloak.migration.migrators.MigrateTo26_4_0.migrate(MigrateTo26_4_0.java:41)
        at org.keycloak.storage.datastore.DefaultMigrationManager.migrate(DefaultMigrationManager.java:165)
        at org.keycloak.migration.MigrationModelManager.migrate(MigrationModelManager.java:33)
        at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.migrateModel(QuarkusJpaConnectionProviderFactory.java:205)
        at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.initSchema(QuarkusJpaConnectionProviderFactory.java:196)
        at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:347)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:461)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:346)
        at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:336)
        at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.postInit(QuarkusJpaConnectionProviderFactory.java:129)
        at org.keycloak.services.DefaultKeycloakSessionFactory.initializeProviders(DefaultKeycloakSessionFactory.java:167)
        at org.keycloak.services.DefaultKeycloakSessionFactory.initProviderFactories(DefaultKeycloakSessionFactory.java:141)
        at org.keycloak.services.DefaultKeycloakSessionFactory.initProviderFactories(DefaultKeycloakSessionFactory.java:125)
        at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.init(QuarkusKeycloakSessionFactory.java:87)
        at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.createSessionFactory(QuarkusKeycloakApplication.java:60)
        at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:75)
        at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.onStartupEvent(QuarkusKeycloakApplication.java:50)
        at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication_Observer_onStartupEvent_VUhusbW4xizn9A5dimKJUcfvjHs.notify(Unknown Source)
        at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:365)
        at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:347)
        at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:81)
        at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:163)
        at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:114)
        at io.quarkus.runner.recorded.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
        at io.quarkus.runner.recorded.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:80)
        at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:139)
        at org.keycloak.quarkus.runtime.cli.Picocli.start(Picocli.java:913)
        at org.keycloak.quarkus.runtime.cli.command.AbstractAutoBuildCommand.runCommand(AbstractAutoBuildCommand.java:134)
        at org.keycloak.quarkus.runtime.cli.command.AbstractCommand.lambda$call$1(AbstractCommand.java:73)
        at java.base/java.util.Optional.orElseGet(Optional.java:364)
        at org.keycloak.quarkus.runtime.cli.command.AbstractCommand.call(AbstractCommand.java:72)
        at org.keycloak.quarkus.runtime.cli.command.AbstractCommand.call(AbstractCommand.java:37)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2031)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2469)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2461)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2423)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2425)
        at picocli.CommandLine.execute(CommandLine.java:2174)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:118)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:109)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:83)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:69)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:37)
Caused by: java.sql.SQLIntegrityConstraintViolationException: Duplicate entry 'uuid://00000000-0000-0000-0000-000000000001' for key 'JGROUPS_PING.PRIMARY'
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:118)
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:912)
        at com.mysql.cj.jdbc.ClientPreparedStatement.execute(ClientPreparedStatement.java:354)
        at io.agroal.pool.wrapper.PreparedStatementWrapper.execute(PreparedStatementWrapper.java:301)
        at org.keycloak.spi.infinispan.impl.embedded.JGroupsConfigurator.lambda$prepareJGroupsAddress$5(JGroupsConfigurator.java:285)
        at org.hibernate.engine.spi.SessionImplementor.lambda$runWithConnection$0(SessionImplementor.java:93)
        at org.hibernate.jdbc.WorkExecutor.executeWork(WorkExecutor.java:37)
        at org.hibernate.internal.AbstractSharedSessionContract.lambda$doWork$8(AbstractSharedSessionContract.java:1151)
        at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.coordinateWork(JdbcCoordinatorImpl.java:314)
        at org.hibernate.internal.AbstractSharedSessionContract.doWork(AbstractSharedSessionContract.java:1162)
        at org.hibernate.internal.AbstractSharedSessionContract.doWork(AbstractSharedSessionContract.java:1150)
        at org.hibernate.engine.spi.SessionImplementor.runWithConnection(SessionImplementor.java:90)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.keycloak.connections.jpa.support.EntityManagerProxy.invoke(EntityManagerProxy.java:100)
        ... 87 more

```

Here is the jdbc_ping table:

<img width="883" height="113" alt="Image" src="https://github.com/user-attachments/assets/ea5b4a76-23fe-4c24-bdb3-8332e9fab7b3" />

My guess is that the pod has a new IP/name after the redployment, but the table was not updated. As I said maybe this is edgy and will never happen if you don´t delete the schema while it´s running, but I just wanted to make clear if this can not happen in other scenarios.

So best would be to start as singleton and delete the table maybe. But here my knowledge ends. 👯 

Here is the Server_Config table, but I think its not the right place to look.

<img width="383" height="113" alt="Image" src="https://github.com/user-attachments/assets/9e1fa0c4-0c35-429a-bc51-9e5d738df03a" />


I think  Iclose it, probably to much of an edge case. I tried a lot and it only happens in this case. 
Caused by: java.sql.SQLIntegrityConstraintViolationException: Duplicate entry 'uuid://00000000-0000-0000-0000-000000000001'

--- a/model/infinispan/src/main/java/org/keycloak/spi/infinispan/impl/embedded/JGroupsConfigurator.java
+++ b/model/infinispan/src/main/java/org/keycloak/spi/infinispan/impl/embedded/JGroupsConfigurator.java
@@ -22,6 +22,7 @@ import java.security.KeyManagementException;
 import java.security.NoSuchAlgorithmException;
 import java.sql.Connection;
 import java.sql.PreparedStatement;
+import java.sql.ResultSet;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -47,6 +48,7 @@ import org.keycloak.jgroups.header.TracerHeader;
 import org.keycloak.jgroups.protocol.KEYCLOAK_JDBC_PING2;
 import org.keycloak.jgroups.protocol.OPEN_TELEMETRY;
 import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
 import org.keycloak.models.utils.KeycloakModelUtils;
 import org.keycloak.provider.ProviderConfigProperty;
 import org.keycloak.provider.ProviderConfigurationBuilder;
@@ -262,22 +264,50 @@ public final class JGroupsConfigurator {
      * one second. This prevents a split-brain scenario on a concurrent startup.
      */
     private static Address prepareJGroupsAddress(KeycloakSession session, String clusterName) {
-        var storage = session.getProvider(ServerConfigStorageProvider.class);
-        String seq = storage.loadOrCreate(JGROUPS_ADDRESS_SEQUENCE, () -> "0");
-        long value = Long.parseLong(seq) + 1;
-        String newSeq = Long.toString(value);
-        storage.replace(JGROUPS_ADDRESS_SEQUENCE, seq, newSeq);
-
         var cp = session.getProvider(JpaConnectionProvider.class);
         var tableName = JpaUtils.getTableNameForNativeQuery("JGROUPS_PING", cp.getEntityManager());
-        String statement = String.format("INSERT INTO %s values (?, ?, ?, ?, ?)", tableName);
 
-        ExtendedUUID address = new ExtendedUUID(0, value);
+        long highestSequence = findHighestSequenceInTable(cp, clusterName, tableName);
+        long mySequence = getNextSequence(session.getKeycloakSessionFactory(), highestSequence);
+        return insertSequenceInTable(cp, clusterName, tableName, mySequence);
+    }
 
+    private static long findHighestSequenceInTable(JpaConnectionProvider cp, String clusterName, String tableName) {
+        return cp.getEntityManager().<Connection, Long>callWithConnection(con -> {
+            long maxSequence = -1;
+            try (PreparedStatement s = con.prepareStatement("SELECT address FROM %s WHERE cluster_name=?".formatted(tableName))) {
+                s.setString(1, clusterName);
+                try (ResultSet resultSet = s.executeQuery()) {
+                    while (resultSet.next()) {
+                        String uuid = resultSet.getString(1);
+                        Address addr = org.jgroups.util.Util.addressFromString(uuid);
+                        if (addr instanceof UUID uuidAddr && uuidAddr.getMostSignificantBits() == 0 && uuidAddr.getLeastSignificantBits() > maxSequence) {
+                            maxSequence = uuidAddr.getLeastSignificantBits();
+                        }
+                    }
+                }
+            }
+            return maxSequence;
+        });
+    }
 
-        cp.getEntityManager().runWithConnection(o -> {
-            Connection con = (Connection) o;
-            try (PreparedStatement s = con.prepareStatement(statement)) {
+    private static long getNextSequence(KeycloakSessionFactory sf, long highestSequence) {
+        // Run this in a separate transaction so that we already write the new ID to the database.
+        // This helps us in case there is an inconsistency if some old restored database and some inconsistent nodes writing
+        // their node IDs to the table. Given this is being called in a retry loop, we'll be making some progress.
+        return KeycloakModelUtils.runJobInTransactionWithResult(sf, session -> {
+            var storage = session.getProvider(ServerConfigStorageProvider.class);
+            String seq = storage.loadOrCreate(JGROUPS_ADDRESS_SEQUENCE, () -> "0");
+            long value = Math.max(highestSequence + 1, Long.parseLong(seq) + 1);
+            storage.replace(JGROUPS_ADDRESS_SEQUENCE, seq, Long.toString(value));
+            return value;
+        });
+    }
+
+    private static Address insertSequenceInTable(JpaConnectionProvider cp, String clusterName, String tableName, long mySequence) {
+        ExtendedUUID address = new ExtendedUUID(0, mySequence);
+        cp.getEntityManager().<Connection>runWithConnection(con -> {
+            try (PreparedStatement s = con.prepareStatement("INSERT INTO %s values (?, ?, ?, ?, ?)".formatted(tableName))) {
                 s.setString(1, org.jgroups.util.Util.addressToString(new UUID(address.getMostSignificantBits(), address.getLeastSignificantBits()))); // address
                 s.setString(2, "(starting)"); // name
                 s.setString(3, clusterName); // cluster name
@@ -286,7 +316,6 @@ public final class JGroupsConfigurator {
                 s.execute();
             }
         });
-
         return address;
     }
 

